storage/google/cloud/storage/blob.py

```diff
@@ -1949,6 +1954,88 @@ def updated(self):
         if value is not None:
             return _rfc3339_to_datetime(value)
 
+    @staticmethod
+    def _get_files(path):
+        """Get the list of full path of all files in the directory.
+        :param path: str
+        :return: The path to the directory.
+
+        :rtype: list
+        :returns: The list of all file path in the directory.
+        """
+        filepath_list = []
+        for directory, _, filenames in os.walk(path):
+            for filename in filenames:
+                filepath_list.append(os.path.join(directory, filename))
+        return filepath_list
+
+    def upload_parallel(
+        self, path, content_type=None, client=None, predefined_acl=None
+    ):
+        """Upload this blob's contents parallel from the content of file in directory.
+
+        The content type of the upload will be determined in order
+        of precedence:
+
+        - The value passed in to this method (if not :data:`None`)
+        - The value stored on the current blob
+        - The value given by ``mimetypes.guess_type``
+        - The default value ('application/octet-stream')
+
+        .. note::
+           The effect of uploading to an existing blob depends on the
+           "versioning" and "lifecycle" policies defined on the blob's
+           bucket.  In the absence of those policies, upload will
+           overwrite any existing contents.
+
+           See the `object versioning
+           <https://cloud.google.com/storage/docs/object-versioning>`_ and
+           `lifecycle <https://cloud.google.com/storage/docs/lifecycle>`_
+           API documents for details.
+
+        If :attr:`user_project` is set on the bucket, bills the API request
+        to that project.
+
+        :type path: str
+        :param path: The path to the directory.
+
+        :type content_type: str
+        :param content_type: Optional type of content being uploaded.
+
+        :type client: :class:`~google.cloud.storage.client.Client`
+        :param client: (Optional) The client to use.  If not passed, falls back
+                       to the ``client`` stored on the blob's bucket.
+
+        :type predefined_acl: str
+        :param predefined_acl: (Optional) predefined access control list
+        """
+        threads = []
+        files_list = self._get_files(path)
+        total_files = len(files_list)
+        max_thread_to_start = min(total_files, _MAX_THREAD_LIMIT)
+        for _ in range(max_thread_to_start):
+            thread = threading.Thread(
+                target=self._upload_from_list,
+                args=(files_list, total_files, content_type, client, predefined_acl),
+            )
```

Comment:
This doesn't seem right. Here you are creating multiple threads that essentially try to upload the same set of files. That is, `_upload_from_list` uses the same `files_list` over and over again. There should be distribution of individual file uploads among separate threads, not duplicating the uploading procedure.

Reply:
I have done the changes which doesn't pass `files_list ` and `total_files` with all threads , but it just passes the `files_list` with all thread as a argument , but `self._files_list[self._file_count] ` line doesn't allow to upload duplicate files because it increment the count with every upload and took the file from list from particular index.

Reply:
How about racing? The index is updated after the upload. Therefore we cannot eliminate the possibility that while a file is being uploaded by a thread, another thread starts uploading the same file. This can also lead to out-of-range exception.